cfg/proc.sql

```diff
@@ -0,0 +1,737 @@
+-- related objects.
+
+\set ON_ERROR_STOP on
+SET client_min_messages = WARNING;
+
+-- rippled error codes. To modify, update the error_label_enum and
+-- the array in the anonymous function just below. PLpg/SQL doesn't have a
+-- tidy way to programmatically insert new enumerated type values.
+DROP TYPE IF EXISTS error_label_enum cascade;
+CREATE TYPE error_label_enum AS ENUM (
+    'rpcINVALID_PARAMS',
+    'rpcNOT_IMPL',
+    'rpcTXN_NOT_FOUND',
+    'rpcINTERNAL',
+    'rpcLGR_IDXS_INVALID',
+    'rpcACT_MALFORMED',
+    'rpcLGR_NOT_FOUND'
+);
+
+-- Attributes of rippled errors.
+DROP TABLE IF EXISTS errors;
+CREATE TABLE IF NOT EXISTS errors (
+    label   error_label_enum,
+    code    int,
+    error   text,
+    message text
+);
+
+-- Anonymous function to populate the errors table. Update the array
+-- in addition to the enum definition to make changes.
+DO $$
+DECLARE
+    _error_list errors[] := ARRAY[
+        ('rpcINVALID_PARAMS', 31, 'invalidParams', 'Invalid parameters.'),
+        ('rpcNOT_IMPL', 74, 'notImpl', 'Not implemented.'),
+        ('rpcTXN_NOT_FOUND', 29, 'txnNotFound', 'Transaction not found.'),
+        ('rpcINTERNAL', 73, 'internal', 'Internal error.'),
+        ('rpcLGR_IDXS_INVALID', 57, 'lgrIdxsInvalid',
+		'Ledger indexes invalid.'),
+        ('rpcACT_MALFORMED', 35, 'actMalformed', 'Account malformed.'),
+        ('rpcLGR_NOT_FOUND', 21, 'lgrNotFound', 'Ledger not found.')
+    ];
+    _error errors;
+BEGIN
+    FOREACH _error IN ARRAY _error_list LOOP
+        INSERT INTO errors VALUES (
+            _error.label,
+            _error.code,
+            _error.error,
+            _error.message
+        );
+    END LOOP;
+END $$;
+
+CREATE UNIQUE INDEX IF NOT EXISTS errors_label_idx on errors (label);
+VACUUM ANALYZE errors;
+
+-- Return json error object based on error type.
+CREATE OR REPLACE FUNCTION return_error (
+    _in_error error_label_enum,
+    _in_error_text text = NULL
+) RETURNS jsonb AS $$
+DECLARE
+    _record record;
+BEGIN
+    EXECUTE 'SELECT * FROM errors WHERE label = $1'
+        INTO _record USING _in_error;
+    RETURN jsonb_build_object('result',
+        jsonb_build_object('status', 'error',
+            'error_message', _record.message,
+            'error', (SELECT CASE WHEN _in_error_text IS NOT NULL THEN
+                _in_error_text ELSE _record.error END),
+            'error_code', _record.code
+        )
+    );
+END;
+$$ LANGUAGE plpgsql;
+
+CREATE OR REPLACE FUNCTION tx (
+    _in_trans_id bytea
+) RETURNS jsonb AS $$
+DECLARE
+    _min_seq           bigint := (SELECT ledger_seq
+	                            FROM ledgers
+				   WHERE ledger_seq = min_ledger()
+			             FOR UPDATE);
+    _max_seq           bigint := max_ledger();
+    _ledger_seq        bigint;
+BEGIN
+    IF _min_seq IS NULL THEN
+        RETURN jsonb_build_object('error', 'empty database');
+    END IF;
+    IF length(_in_trans_id) != 20 THEN
```

Comment:
the transaction hashes are 32 bytes actually

Reply:
fixed